lib-ee/emqx_license/test/emqx_license_SUITE.erl

```diff
@@ -0,0 +1,133 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_license_SUITE).
+
+-compile(nowarn_export_all).
+-compile(export_all).
+
+-include_lib("emqx/include/emqx_mqtt.hrl").
+
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("common_test/include/ct.hrl").
+
+all() ->
+    emqx_common_test_helpers:all(?MODULE).
+
+init_per_suite(Config) ->
+    _ = application:load(emqx_conf),
+    emqx_config:save_schema_mod_and_names(emqx_license_schema),
+    emqx_common_test_helpers:start_apps([emqx_license], fun set_special_configs/1),
+    Config.
+
+end_per_suite(_) ->
+    emqx_common_test_helpers:stop_apps([emqx_license]),
+    ok.
+
+init_per_testcase(_Case, Config) ->
+    {ok, _} = emqx_cluster_rpc:start_link(node(), emqx_cluster_rpc, 1000),
+    meck:new(emqx_license_parser, [passthrough]),
+    meck:expect(emqx_license_parser, public_key, fun public_key/0),
+    meck:expect(emqx_license_parser, digest, fun digest/0),
+    Config.
+
+end_per_testcase(_Case, _Config) ->
+    meck:unload(emqx_license_parser),
+    ok.
+
+set_special_configs(emqx_license) ->
+    Config = #{type => file,
+               value => emqx_license_test_lib:default_license()},
+    emqx_config:put([license], Config),
+    RawConfig = #{<<"type">> => <<"file">>,
+                  <<"value">> => emqx_license_test_lib:default_license()},
+    emqx_config:put_raw([<<"license">>], RawConfig);
+
+set_special_configs(_) -> ok.
+
+%%------------------------------------------------------------------------------
+%% Tests
+%%------------------------------------------------------------------------------
+
+t_update_file(_Config) ->
+    ?assertMatch(
+       {error, {invalid_license_file, enoent}},
+       emqx_license:update_file("/unknown/path")),
+
+    ?assertMatch(
+       {ok, #{}},
+       emqx_license:update_file(emqx_license_test_lib:default_license())).
+
+t_update_value(_Config) ->
+    ?assertMatch(
+       {error, {unknown_format, _}},
+       emqx_license:update_value("invalid.license")),
+
+    {ok, LicenseValue} = file:read_file(emqx_license_test_lib:default_license()),
+
+    ?assertMatch(
+       {ok, #{}},
+       emqx_license:update_value(LicenseValue)).
+
+t_check_exceeded(_Config) ->
+    License = mk_license(
+                ["220111",
+                 "0",
+                 "10",
+                 "Foo",
+                 "contact@foo.com",
+                 "20220111",
+                 "100000",
+                 "10"]),
+    #{} = emqx_license_checker:update(License),
+
+    ok = lists:foreach(
+           fun(_) ->
+                   {ok, C} = emqtt:start_link(),
+                   {ok, _} = emqtt:connect(C)
+           end,
+           lists:seq(1, 11)),
+
+    ?assertEqual(
+       {stop, {error, ?RC_QUOTA_EXCEEDED}},
+       emqx_license:check(#{}, #{})).
```

Comment:
Q: if we allow 10 % more clients than the configured maximum, shouldn't it be ok to have 11 clients if the maximum is 10?  :thinking:

Reply:
This is because of non-strict inequality check in the callback hook.
I changed it to strict so that it is easier to estimate number of allowed connections.